[AArch64][LV][SLP] Vectorizers use call cost for vectorized frem

paschalis-mpeis opened the pull request:
SLP vectorization for frem now happens when vector library calls are
available, given its type and vector length. This is due to using the
updated cost that amounts to a call.
    
Add tests that do SLP vectorization for code that contains 2x double and
4x float frem instructions.
    
LoopVectorizer now also uses getFRemInstrCost.

paschalis-mpeis committed:
[AArch64] SLP can vectorize frem

When vector library calls are available for frem, given its type and
vector length, the SLP vectorizer uses updated costs that amount to a
call, matching LoopVectorizer's functionality.

This allows 'superword-level' vectorization, which can be converted to
a vector lib call by later passes.

Add tests that vectorize code that contains 2x double and 4x float frem
instructions.

paschalis-mpeis committed:
SLP cannot vectorize frem calls in AArch64.

It needs updated costs when there are available vector library functions
given the VF and type.

alexey-bataev commented on llvm/lib/Transforms/Vectorize/SLPVectorizer.cpp:
```diff
@@ -8362,9 +8362,20 @@ BoUpSLP::getEntryCost(const TreeEntry *E, ArrayRef<Value *> VectorizedVals,
-      return TTI->getArithmeticInstrCost(ShuffleOrOp, VecTy, CostKind, Op1Info,
-                                         Op2Info) +
-             CommonCost;
+      auto VecCost = TTI->getArithmeticInstrCost(ShuffleOrOp, VecTy, CostKind,
+                                                 Op1Info, Op2Info);
+      // Some targets can replace frem with vector library calls.
+      if (ShuffleOrOp == Instruction::FRem) {
+        LibFunc Func;
+        if (TLI->getLibFunc(ShuffleOrOp, ScalarTy, Func) &&
+            TLI->isFunctionVectorizable(TLI->getName(Func),
+                                        VecTy->getElementCount())) {
+          auto VecCallCost = TTI->getCallInstrCost(
+              nullptr, VecTy, {ScalarTy, ScalarTy}, CostKind);
+          VecCost = std::min(VecCost, VecCallCost);
+        }
+      }
+      return VecCost + CommonCost;
```
This should be a seaprate patch.

alexey-bataev commented on llvm/lib/Transforms/Vectorize/SLPVectorizer.cpp:
```diff
@@ -8362,9 +8362,20 @@ BoUpSLP::getEntryCost(const TreeEntry *E, ArrayRef<Value *> VectorizedVals,
-      return TTI->getArithmeticInstrCost(ShuffleOrOp, VecTy, CostKind, Op1Info,
-                                         Op2Info) +
-             CommonCost;
+      auto VecCost = TTI->getArithmeticInstrCost(ShuffleOrOp, VecTy, CostKind,
```
Do not use auto

alexey-bataev commented on llvm/lib/Transforms/Vectorize/SLPVectorizer.cpp:
```diff
@@ -8362,9 +8362,20 @@ BoUpSLP::getEntryCost(const TreeEntry *E, ArrayRef<Value *> VectorizedVals,
+        if (TLI->getLibFunc(ShuffleOrOp, ScalarTy, Func) &&
+            TLI->isFunctionVectorizable(TLI->getName(Func),
+                                        VecTy->getElementCount())) {
+          auto VecCallCost = TTI->getCallInstrCost(
```
Do not use auto

alexey-bataev commented on llvm/lib/Transforms/Vectorize/SLPVectorizer.cpp:
```diff
@@ -8362,9 +8362,20 @@ BoUpSLP::getEntryCost(const TreeEntry *E, ArrayRef<Value *> VectorizedVals,
+      // Some targets can replace frem with vector library calls.
+      if (ShuffleOrOp == Instruction::FRem) {
+        LibFunc Func;
+        if (TLI->getLibFunc(ShuffleOrOp, ScalarTy, Func) &&
+            TLI->isFunctionVectorizable(TLI->getName(Func),
+                                        VecTy->getElementCount())) {
```
This should be hidden in TTI, vectorizers use TTI interfaces. Encapsulate it into getArithmeticInstrCost, then you won't need to change SLP vectorizer

paschalis-mpeis replied:
I'd rather not adapt `TTI.getArithmeticInstrCost` API, as that would cause changes in 150+ places, including every target that overrides that. But I understand your concern as this is specific, affecting one of our instructions.

TLI is simply passed to `getVecLibCallCost`, just like it's widely passed around in other places in the SLP code. This isn't making the code any target dependent, TLI/TTI are abstractions to that.

Internally, `getVecLibCallCost` does those checks, which have now moved to `VectorUtils`.
SLP gets a valid cost if there's one, and picks the minimum between the two.

huntergr-arm commented on llvm/lib/Target/AArch64/AArch64TargetTransformInfo.cpp:
```diff
@@ -2972,6 +2972,13 @@ InstructionCost AArch64TTIImpl::getArithmeticInstrCost(
                                          Op2Info);
+  case ISD::FREM:
+    // Pass nullptr as fmod/fmodf calls are emitted by the backend even when
+    // those functions are not delcared in the module.
```
nit: 'declared'

paschalis-mpeis commented:
Addressed reviewers and rebased to parent pr:
-  #80423

Github is now rendering **only** the changes of this patch.

paulwalker-arm commented on llvm/lib/Analysis/TargetTransformInfo.cpp:
```diff
@@ -869,6 +870,18 @@ TargetTransformInfo::getOperandInfo(const Value *V) {
+    TTI::TargetCostKind CostKind) {
+  Type *ScalarTy = VecTy->getScalarType();
+  LibFunc Func;
+  if (TLI->getLibFunc(OpCode, ScalarTy, Func) &&
+      TLI->isFunctionVectorizable(TLI->getName(Func), VecTy->getElementCount()))
```
This seems to be surreptitiously adding another mechanism to check for the presence of a vector math routine. Under what circumstance do you need to check for the cost of something that might not exist? I would expect TLI to be queried directly as part of a transformation and once they've concluded a vector math call exists then they'd simply query the cost of the call.

paschalis-mpeis replied:
Actually, in LoopVectorizer we have done it this way.

For the sake of reusing the code between the 2 vectorizers, I've put it in TTI, but after our discussion I've now moved it to `VectorUtils`. LoopVectorizer should now reuse this method.

alexey-bataev replied:
I disagree with this approach and still insist that it should be implemented in TTI rather doing some target specific "magic" in the vectorizers

paulwalker-arm replied:
TTI should be costing known entities and not trying to second guess what a transformation is doing.  Whilst TLI provides a route for target specific queries relating to the availability of vector math routines, that doesn’t mean those mapping will be used.  The only source of truth is the transformation pass itself and thus it needs to ask the TTI the correct question based on this source of truth (i.e. whether it intends to use TLI mappings[1]). Both TTI and TLI abstract away the target specific queries relating to their design so I don’t understand what is the target specific “magic” you are worried about.

[1] This is very important because if the vector function is not used then a scalable vector FREM must have an invalid cost because there is no code generation support for it.

alexey-bataev replied:
It does not known what transsformation is doing, it does exactly what you said - estimates the cost of the target-specific code. WE're not emitting frem calls directly, we're still emitting regular frem vector code, which is then lowered specifically for ARM. And target specific TTI should be aware of this.

paulwalker-arm replied:
OK, I see what you're saying.  I guess the problem here is that we're relying on a subsequent pass (ReplaceWithVecLib) to be present in the pipeline in order to correctly cost the FREM. With neither SLP nor TTI having this knowledge?

alexey-bataev replied:
Yes, and what I'm saying that TTI should handle this since this is target-specific lowering.
If you don't want to propagate TLI, then just copy the legality checks from TLI to TTI and the cost estimation for this specific lowering to the getArithmeticInstructionCost function. This is what we usually do in similar cases for other targets.

paulwalker-arm commented on llvm/lib/Analysis/TargetTransformInfo.cpp:
```diff
@@ -869,6 +870,18 @@ TargetTransformInfo::getOperandInfo(const Value *V) {
+  LibFunc Func;
+  if (TLI->getLibFunc(OpCode, ScalarTy, Func) &&
+      TLI->isFunctionVectorizable(TLI->getName(Func), VecTy->getElementCount()))
+    return getCallInstrCost(nullptr, VecTy, {ScalarTy, ScalarTy}, CostKind);
```
This is hardwiring two scalar parameters when costing the call?

paschalis-mpeis committed:
Addressing reviewers

paulwalker-arm reviewed (commented):
I've review the patch from both side so most of the comment will be void if you opt for the new TTI hook.  That advantage of the TTI hook is that because it is specific to FREM you can hardware things like numbers of operands, which should streamline the implementation.

alexey-bataev commented:
> Personally I'm happy with keeping this nuance outside of TTI but if we really want this captured within TTI then I think it's time to break FREM into its own cost function (i.e. implement getFRemInstrCost. That way getArithmeticInstrCost can work as it does today and the new function can be documented to highlight it's assumption that if a TLI is passed in and a vector mapping is present then the return value is only valid based on it's assumption that vector FREM instructions will be transformed by a following transformation pass. I prefer this to say, adding TLI to getArithmeticInstrCost, because I'd rather users of getFRemInstrCost to explicitly enter into this contract.

Hm, not sure adding getFRemInstrCost is the best solution here. I would more support adding TLI to getArithmeticInstrCost instead. Some other users may benefit from this too. Though getFRemInstrCost is still better than the current solution

paulwalker-arm commented:
Changing `getArithmeticInstrCost` is just too dangerous.  What if one opcode needs TLI for a different reason? all of a sudden all existing callers are entered into the contract (FREM is guaranteed to be transformed into a math call) without ensuring that's actually the case.

alexey-bataev commented:
> Changing `getArithmeticInstrCost` is just too dangerous. What if one opcode needs TLI for a different reason?

That should be fine, what's the dangerous in it?

paschalis-mpeis commented:
The benefits of having`getFRemInstrCost` in my view are the below:
1. frem is a special case anyway:
It's an IR instruction that is not supported by all hw and targets have to specialize.
Handling it in a dedicated switch case with a dedicated TTI function call, clearly exposes that information to anyone who reads the code in both vectorizers (and not obscuring it away). 
Plus it won't add any `if (TLI hasVecLib) doThis else doThat` logic to the vectorizers.

2. This won't be a significant API change.
It won't force any other user of the `getArithmeticInstrCost` to go through that change.

paschalis-mpeis committed:
[LV][SLP] Vectorizers now use getFRemInstrCost for frem costs

SLP vectorization for frem now happens when vector library calls are
available, given its type and vector length. This is due to using the
updated cost that amounts to a call.

Add tests that do SLP vectorization for code that contains 2x double and
4x float frem instructions.

LoopVectorizer now also uses getFRemInstrCost.

paschalis-mpeis commented:
forced-push to rebase to main (parent PR #80423 merged, so this one is no longer stacked).

This change, matches functionality on both LoopVectorizer and SLPVectorizer.

What do you think about the latest change?
Thank you all for the valuable feedback so far.

paulwalker-arm commented on llvm/lib/Analysis/TargetTransformInfo.cpp:
```diff
@@ -881,6 +882,24 @@ InstructionCost TargetTransformInfo::getArithmeticInstrCost(
+  assert(Opcode == Instruction::FRem && "Instruction must be frem");
+
+  VectorType *VecTy = dyn_cast<VectorType>(Ty);
+  Type *ScalarTy = VecTy ? VecTy->getScalarType() : Ty;
```
`getScalarType()` has the correct behaviour for vectors and scalars so you can confidently use `Ty-> getScalarType()` as part of `getLibFunc` call.

paulwalker-arm commented on llvm/lib/Transforms/Vectorize/LoopVectorize.cpp:
```diff
@@ -6899,25 +6899,15 @@ LoopVectorizationCostModel::getInstructionCost(Instruction *I, ElementCount VF,
-    return std::min(InstrCost, VecCallCost);
+    TTI::OperandValueInfo Op1Info{TargetTransformInfo::OK_AnyValue,
+                                  TargetTransformInfo::OP_None};
+    // Some targets replace frem with vector library calls.
```
This comment is not longer relevant.  We're simply using a more specific costing function, which is clear from the code.

paulwalker-arm commented on llvm/lib/Transforms/Vectorize/SLPVectorizer.cpp:
```diff
@@ -8616,6 +8616,13 @@ BoUpSLP::getEntryCost(const TreeEntry *E, ArrayRef<Value *> VectorizedVals,
       TTI::OperandValueInfo Op1Info = getOperandInfo(E->getOperand(0));
       TTI::OperandValueInfo Op2Info = getOperandInfo(E->getOperand(OpIdx));
+
+      // Some targets replace frem with vector library calls.
```
As above, the comment is unnecessary.

paulwalker-arm commented on llvm/include/llvm/Analysis/TargetTransformInfo.h:
```diff
@@ -1255,6 +1255,18 @@ class TargetTransformInfo {
+  /// Returns the cost of a vector instruction based on the assumption that frem
+  /// will be later transformed (by ReplaceWithVecLib) into a call to a
+  /// platform specific frem vector math function.
+  /// If unsupported, it will return cost using getArithmeticInstrCost.
```
Perhaps "Returns the same cost as getArithmeticInstrCost when no math function is available." ?

paulwalker-arm reviewed (commented):
A few minor comments but others this looks fine to me.

paschalis-mpeis committed:
Addressing reviewers (2)

alexey-bataev commented:
> The benefits of having`getFRemInstrCost` in my view are the below:
> 
> 1. frem is a special case anyway:
>    It's an IR instruction that is not supported by all hw and targets have to specialize.
>    Handling it ~in a dedicated switch case~ with a dedicated TTI function call, clearly exposes that information to anyone who reads the code in both vectorizers (and not obscuring it away).

What's so special about it? It is absolutely fine if the target does not support it.

>    Plus it won't add any `if (TLI hasVecLib) doThis else doThat` logic to the vectorizers.
> 2. This won't be a significant API change.
>    It won't force any other user of the `getArithmeticInstrCost` to go through that change.
> 

This is not a problem at all. Instead you're adding a special API for a single instruction, which bloats the users of this API.

> Edit: I won't use a dedicate switch as I don't want to duplicate those lambdas or introduce any weird fall-throughs.

paschalis-mpeis commented:
> Alexey: That should be fine, what's the dangerous in it?

Currently,`frem` relies on `ReplaceWithVeclib` pass, so if for whatever reason that pass does not ran, then codegen would crash.


> Paul: What if one opcode needs TLI for a different reason? all of a sudden all existing callers are entered into the contract (FREM is guaranteed to be transformed into a math call) without ensuring that's actually the case

If we encapsulate this functionality in `getArithmeticInstrCost`, we are adding existing callers into this 'contract'.
What this means, is that they could potentially pass at some point in the future a TLI object, returning the updated costs in a scenario where `ReplaceWithVeclib` happens to **not** ran.

So, until CodeGen gains support for emitting vectorized frem, we want to be very explicit about it.

### Examples:
1) A new pass might calls `getArithmeticInstrCost` and passes TLI there, which may happen to have veclib support for frem. 

2) Similarly, a maintaiiner might decide to pass TLI to an existing callsite of `getArithmeticInstrCost`, getting again the updated costs.

In both examples, the return costs will favor replacing to a vectorized frem, however, it could be at a point in the pipeline where `ReplaceWithVeclib` does not ran, causing codegen to crash soon after.


---

## What's next:
Ideally, we would also prefer to have this merged into `getArithmeticInstrCost`, but we are reluctant to do so until Codegen learns to deal with `frem` in all cases.

So until then, we insist on being very explicit on those cases that knowingly call the special `getFRemInstrCost` (SLP & Loop vectorizers). At this time, I could add a `TODO` note of such an intention to `getFRemInstrCost`, and in the future, given such codegen support, we'd be happy to fully abstract it away inside TTI.

@alexey-bataev  while not ideal, does the above make things clearer? Also, given we proceed with this approach for this patch, what changes would you require?

alexey-bataev commented:
> Currently,`frem` relies on `ReplaceWithVeclib` pass, so if for whatever reason that pass does not ran, then codegen would crash.

Same applies to the newly introduced function

> If we encapsulate this functionality in `getArithmeticInstrCost`, we are adding existing callers into this 'contract'.
> What this means, is that they could potentially pass at some point in the future a TLI object, returning the updated costs in a scenario where `ReplaceWithVeclib` happens to **not** ran.

This is not the problem of this patch, it is the issue of future patches and the authors of these patches should handle it correctly.

paschalis-mpeis commented:
> Same applies to the newly introduced function

True, but the newly introduced function makes it quite explicit to it's users.
They can't use it by mistake without noticing. And now we'll have it in two places that we know are OK.

> This is not the problem of this patch, it is the issue of future patches and the authors of these patches should handle it correctly.

Exactly. Until codegen can fully handle frem, we'd prefer not to set any 'traps' for future patches and authors, and instead be very explicit about it.

alexey-bataev commented:
> True, but the newly introduced function makes it quite explicit to it's users.
> They can't use it by mistake without noticing. And now we'll have it in two places that we know are OK.

Same with adding the new parameter to the existing function, no difference at all.

> Exactly. Until codegen can fully handle frem, we'd prefer not to set any 'traps' for future patches and authors, and instead be very explicit about it.

As I said, this is not a problem of this patch

paulwalker-arm commented:
If adding this new function that behaves differently to `getArithmeticInstrCost` for the reasons described is such a sticking point then I think we cannot safely use scalable vector `FRem` instructions on AArch64 and should instead change `LoopVectorize` and `SLPVectorize` to cost and emit the function calls directly.  Ideally I’d rather not complicate those passes but it would better reflect the transformation that is actually happening and thus remove my concerns and any need to modify TTI.

paulwalker-arm commented:
Perhaps a better option is for the code generater to emit the function calls as part of legalisation.  If we did that then I'd be more comfortable with modifying `getArithmeticInstrCost`.

paulwalker-arm commented:
@paschalis-mpeis, I'll investigate potential code generator changes and will report back.

RKSimon commented on llvm/include/llvm/Analysis/TargetTransformInfo.h:
```diff
@@ -1255,6 +1255,19 @@ class TargetTransformInfo {
+  /// platform specific frem vector math function.
+  /// Returns the same cost as getArithmeticInstrCost when no math function is
+  /// available.
+  InstructionCost getFRemInstrCost(
```
Might it not be better to make this another getArithmeticInstrCost variant that can fall back to the default getArithmeticInstrCost implementation? That way we have less work to when we merge these back together (there will be other opcodes that benefit from LibInfo handling - int div etc.)

paschalis-mpeis committed:
[AArch64][LV][SLP] Vectorizers use call cost for vectorized frem

getArithmeticInstrCost is used by both LoopVectorizer and SLPVectorizer
to compute the cost of frem, which becomes a call cost on AArch64 when
TLI has a vector library function.

Add tests that do SLP vectorization for code that contains 2x double and
4x float frem instructions.

alexey-bataev commented on llvm/lib/Transforms/Vectorize/SLPVectorizer.cpp:
```diff
@@ -8852,7 +8852,8 @@ BoUpSLP::getEntryCost(const TreeEntry *E, ArrayRef<Value *> VectorizedVals,
       TTI::OperandValueInfo Op2Info = getOperandInfo(E->getOperand(OpIdx));
       return TTI->getArithmeticInstrCost(ShuffleOrOp, VecTy, CostKind, Op1Info,
-                                         Op2Info) +
+                                         Op2Info, ArrayRef<const Value *>(),
```
```suggestion
                                         Op2Info, std::nullopt,
```

paschalis-mpeis replied:
Applied in the latest commit: (97a2ad9fb87cbb53328936b88341dbd15bde3585).
Thanks for the suggestion and the comments throughout the reviewing process.

paulwalker-arm commented:
I've landed https://github.com/llvm/llvm-project/pull/83859 so my previous concerns/objections are no longer relevant and have reset my approval accordingly.

paschalis-mpeis commented:
The below patch has landed by @paulwalker-arm allows us to safety encapsulate all functionality in `getArithmeticInstrCost`
- #83859

This force-update rebases to main and restructure the code in SLP and Loop Vectorizers.

paschalis-mpeis committed:
Addressing reviewers (3)

paulwalker-arm reviewed (approved):
I've a couple of suggestions to consider but otherwise looks good.

paschalis-mpeis committed:
Addressing reviewers (4)

alexey-bataev reviewed (approved):
LG
